src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1365,6 +1365,28 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         tmp = ASR::make_ErrorStop_t(al, x.base.base.loc, code);
     }
 
+    void visit_Set(const AST::Set_t &x) {
+        LFORTRAN_ASSERT(x.n_elts > 0); // type({}) is 'dict'
+        Vec<ASR::expr_t*> elts;
+        elts.reserve(al, x.n_elts);
+        ASR::ttype_t* type = nullptr;
+        for (size_t i = 0; i < x.n_elts; ++i) {
+            visit_expr(*x.m_elts[i]);
+            ASR::expr_t *value = ASRUtils::EXPR(tmp);
+            if (type == nullptr) {
+                type = ASRUtils::expr_type(value);
+            } else {
+                if (!ASRUtils::check_equal_type(ASRUtils::expr_type(value), type)) {
+                    throw SemanticError("All Set values must be of the same type for now",
+                                        x.base.base.loc);
+                }
+            }
+            elts.push_back(al, value);
+        }
+
+        tmp = ASR::make_ConstantSet_t(al, x.base.base.loc, elts.p, elts.size(), type);
```

Comment:
The `type` is the type of the elements, and it is correct that you check that they are all of the same type.

However, the type of `ConstantSet` is not the type of the element (say `i32`), but rather a new type `Set<i32>`. So we have to extend our types with a new type `Set`, and that type has to have a member of a type of the element. Then we need to make sure all backends and tests work, so we really need to do this in LFortran to ensure everything works with the new type(s). While at it, we should also add a type Dict.

Reply:
> The `type` is the type of the elements, and it is correct that you check that they are all of the same types.

Down the road, we ought to make support for different types of elements inside a set, right?
```python
>>> {1, -2, -3.4, "a"}
{1, 'a', -3.4, -2}
```

Reply:
It's the same issue as with a list of different elements. I wouldn't worry about it for now, for performance reasons I would focus on the same element type.